test/integration/ws/common.dart

```diff
@@ -225,61 +227,169 @@ void runCommonWebSocketIntegrationTests({connect(Uri uri), int port}) {
     await socket.done;
     expect(socket.closeCode, equals(4001));
     expect(socket.closeReason, equals('Closed by server.'));
   });
+
+  test('message events should be discarded prior to a subscription', () async {
+    var webSocket = await connect(echoUri);
+
+    webSocket.add('1');
+    webSocket.add('2');
+    await new Future.delayed(new Duration(milliseconds: 500));
+
+    var messages = [];
+    webSocket.listen((data) {
+      messages.add(data);
+    });
+
+    webSocket.add('3');
+    webSocket.add('4');
+    await new Future.delayed(new Duration(milliseconds: 500));
+
+    await webSocket.close();
+    expect(messages, orderedEquals(['3', '4']));
+  });
+
+  test(
+      'the first event should be received if a subscription is made immediately',
+      () async {
+    var webSocket = await connect(echoUri);
+
+    var c = new Completer();
+    webSocket.listen((data) {
+      c.complete(data);
+    });
+    webSocket.add('first');
+
+    expect(await c.future, equals('first'));
+    await webSocket.close();
+  });
+
+  test('all event streams should respect pause() and resume() signals',
+      () async {
+    var webSocket = await connect(echoUri);
+    var messages = [];
+
+    // no subscription yet, messages should be discarded
+    webSocket.add('1');
+    await new Future.delayed(new Duration(milliseconds: 500));
+
+    // setup a subscription, messages should be recorded
+    var sub = webSocket.listen((data) {
+      messages.add(data);
+    });
+    webSocket.add('2');
+    await new Future.delayed(new Duration(milliseconds: 500));
+
+    // pause the subscription, messages should be discarded
+    sub.pause();
+    await new Future.delayed(new Duration(milliseconds: 500));
+    webSocket.add('3');
+    await new Future.delayed(new Duration(milliseconds: 500));
+
+    // resume the subscription, messages should be recorded again
+    sub.resume();
+    await new Future.delayed(new Duration(milliseconds: 500));
+    webSocket.add('4');
+    await new Future.delayed(new Duration(milliseconds: 500));
+
+    expect(messages, orderedEquals(['2', '4']));
+    await webSocket.close();
+  });
+
+  test('should support reassigning the onData() handler', () async {
+    var webSocket = await connect(echoUri);
+
+    var origMessages = [];
+    var origOnData = (data) {
+      origMessages.add(data);
+    };
+
+    var newMessages = [];
+    var newOnData = (data) {
+      newMessages.add(data);
+    };
+
+    var subscription = webSocket.listen(origOnData);
+    webSocket.add('1');
+    webSocket.add('2');
+    // SockJS requires a delay longer than 1 tick for the echos to be received.
+    await new Future.delayed(new Duration(milliseconds: 500));
+
+    subscription.onData(newOnData);
+    webSocket.add('3');
+    webSocket.add('4');
+    // SockJS requires a delay longer than 1 tick for the echos to be received.
+    await new Future.delayed(new Duration(milliseconds: 500));
```

Comment:
It seemed that the SockJS integration tests needed a fairly large delay, but it may have been related to the socket connections not being closed/cleaned up at the end of every test, so I'll try lowering these delays and removing some that may be unnecessary.